lib/surface/compiler.ex

```diff
@@ -515,11 +515,37 @@ defmodule Surface.Compiler do
 
   defp process_attributes(_module, [], _meta), do: []
 
-  defp process_attributes(mod, [{name, value, attr_meta} | attrs], meta) do
+  defp process_attributes(mod, attrs, meta), do: process_attributes(mod, attrs, meta, [])
+
+  defp process_attributes(_module, [], _meta, acc) do
+    acc
+    |> Keyword.values()
+    |> Enum.reverse()
+  end
+
+  defp process_attributes(mod, [{name, value, attr_meta} | attrs], meta, acc) do
     name = String.to_atom(name)
     attr_meta = Helpers.to_meta(attr_meta, meta)
     {type, type_opts} = Surface.TypeHandler.attribute_type_and_opts(mod, name, attr_meta)
 
+    accumulate? = Keyword.get(type_opts, :accumulate, false)
+
+    if not accumulate? and Keyword.has_key?(acc, name) do
```

Comment:
I'm a bit concern about `Keyword.has_key?(acc, name)`. Since most of the attributes are not accumulative, this will run for almost every attribute, traversing the remaining attributes multiple times (unlike maps, keywords must be fully traversed to check for a key). We have this kind of logic in many other places but we need to start avoiding this so we don't make compilation slower (it's already a bit slower than it should, IMO). 

My suggestion is to instead of doing this validation on every `process_attributes` call, we move this logic to a single place (maybe `validate_properties`?). There we can do a `reduce` or use a `group_by` to process the validation. This way we guarantee we traverse the list only twice. One for grouping/counting, another one to emit the warnings. 

In the future, we should consider having a single optimized data structure built by a `reduce`, gathering all kinds of information in sets and maps, so we can pass it and use it efficiently in multiple processings/validations.

If you prefer, feel free to keep the current implementation. We can search for all similar cases and create another PR with the optimizations.

Reply:
> unlike maps, keywords must be fully traversed to check for a key

I didn't know that, I really though it was like a map. Neither the Elixir nor the Erlang documentation mention that.

I have done benchmarks and there is no increase in time no matter how big the list is if you search for an item in the same place.

> In the future, we should consider having a single optimized data structure built by a single reduce, gathering all kinds of information in sets and maps, so we can pass it and use it efficiently in multiple processings/validations.

I agree.

> If you prefer, feel free to keep the current implementation. We can search for all similar cases and create another PR with the optimizations.

My first thought is to do the right thing right away but I wonder if it wouldn't be better to put this in a pr with the other optimsiations.

Reply:
> Your comment about the complexity is totally right! A Map as a complexity of O(1) and a list of O(n) to check if a key exists. I was challenging the part that says we have to traverse all the list to check for a key. In the worst case this is right.